[RISCV] Use zext.[b/w/h] instead of th.extu for bitfield extraction

Joshua-xt opened the pull request:
We prefer not to do the extu optimization in https://github.com/llvm/llvm-project/commit/10b7cd660c73f704ad8610828e80b33c241451df if we can directly use one zext.b/zext.h/zext.w instruction with zba/zbb extension, since we always give preference to standard extension instead of vendor extension.

topperc commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -1228,6 +1228,10 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
       if (!Subtarget->hasVendorXTHeadBb())
         return false;
 
+      if ((Subtarget->hasStdExtZba() || Subtarget->hasStdExtZbb()) &&
```
Should we be checking bit 31 for hasStdExtZba and bit 7 and 15 for hasStdExtZbb?

joshua-arch1 replied:
Done.

dtcxzyw commented:
Is there a xuantie core that supports both zbb and xtheadbb?

ptomsich commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -1228,6 +1228,10 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
 
+      if ((Subtarget->hasStdExtZba() || Subtarget->hasStdExtZbb()) &&
+          (Msb == 7 || Msb == 15 || Msb == 31) && Lsb == 0)
+        return false;
```
This, of course, assumes that Zba/Zbb instructions have the same or lower cost than XTheadBB instructions, if implemented on the same core.

joshua-arch1 commented:
> Is there a xuantie core that supports both zbb and xtheadbb?

Yeah, I have doone some experiments on XuanTie C908 processors, extu rd,rs,15,0 performs worse than zext.h rd,rs and extu rd,rs,31,0 performs worse than zext.w rd,rs.

Joshua-xt committed:
[RISCV] Use zext.[b/h/w] instead of th.extu for bitfield extraction

We prefer not to do the extu optimization in https://github.com/llvm/
llvm-project/commit/10b7cd660c73f704ad8610828e80b33c241451df if we
can directly use one zext.b/zext.h/zext.w instruction with zba/zbb
extension, since we always give preference to standard extension
instead of vendor extension.

ptomsich commented:
> > Is there a xuantie core that supports both zbb and xtheadbb?
> 
> Yeah, I have done some experiments on XuanTie C908 processors, extu rd,rs,15,0 performs worse than zext.h rd,rs and extu rd,rs,31,0 performs worse than zext.w rd,rs.

I would have expected that both forms decode to the same u-ops
Looks like those are different implementations in the core.

That said, I would recommend we use a new `Feature...` definition and test to see which code-generation strategy a given code prefers.

topperc commented:
> > > Is there a xuantie core that supports both zbb and xtheadbb?
> > 
> > 
> > Yeah, I have done some experiments on XuanTie C908 processors, extu rd,rs,15,0 performs worse than zext.h rd,rs and extu rd,rs,31,0 performs worse than zext.w rd,rs.
> 
> I would have expected that both forms decode to the same u-ops Looks like those are different implementations in the core.
> 
> That said, I would recommend we use a new `Feature...` definition and test to see which code-generation strategy a given code prefers.

How would the new `Feature...` get set? If its not connected to a specific -mcpu/-mtune then user would have to write something like `-Xclang -target-feature -Xclang +new-feature`.

joshua-arch1 commented:
> > > > Is there a xuantie core that supports both zbb and xtheadbb?
> > > 
> > > 
> > > Yeah, I have done some experiments on XuanTie C908 processors, extu rd,rs,15,0 performs worse than zext.h rd,rs and extu rd,rs,31,0 performs worse than zext.w rd,rs.
> > 
> > 
> > I would have expected that both forms decode to the same u-ops Looks like those are different implementations in the core.
> > That said, I would recommend we use a new `Feature...` definition and test to see which code-generation strategy a given code prefers.
> 
> How would the new `Feature...` get set? If its not connected to a specific -mcpu/-mtune then user would have to write something like `-Xclang -target-feature -Xclang +new-feature`.

Yeah, I have the same question. Also, even if we use the feature, we need to have a default code generation. From my perspective, I prefer to use the standard instruction by default.

joshua-arch1 commented:
> > > > Is there a xuantie core that supports both zbb and xtheadbb?
> > > 
> > > 
> > > Yeah, I have done some experiments on XuanTie C908 processors, extu rd,rs,15,0 performs worse than zext.h rd,rs and extu rd,rs,31,0 performs worse than zext.w rd,rs.
> > 
> > 
> > I would have expected that both forms decode to the same u-ops Looks like those are different implementations in the core.
> > That said, I would recommend we use a new `Feature...` definition and test to see which code-generation strategy a given code prefers.
> 
> How would the new `Feature...` get set? If its not connected to a specific -mcpu/-mtune then user would have to write something like `-Xclang -target-feature -Xclang +new-feature`.

Yeah, I have the same question. Also, even if we use the feature, we need to have a default code generation. From my perspective, I prefer to use the standard instruction by default.
